### PR TITLE
DOCK-2507: Add "Discover Existing Tools/Notebooks"

### PR DESCRIPTION
**Pull request by ll5zh**
**Description**
Previously, we discovered that the "Discover Existing Notebooks" button was missing from the UI. Also, "Discover Existing Tools" (disabled in https://github.com/dockstore/dockstore-ui2/pull/1919) wasn't behaving as expected: when clicked, bioworkflows were displayed instead of tools.

This PR should fix/add the functionality for tools and notebooks respectively. To display the correct entries, the UI should now [re-request entries of the desired type](https://ucsc-cgl.atlassian.net/browse/SEAB-6186?focusedCommentId=46748) after calling `addUserToDockstoreWorkflows` (which appears to only return bioworkflows).

_"Discover Existing Tools" buttons and tooltip (restored the HTML that was removed in https://github.com/dockstore/dockstore-ui2/pull/1919):_
Screenshot from 2024-04-01 11-18-11

_"Discover Existing Notebooks" buttons and tooltip:_
Screenshot from 2024-04-01 11-18-21

**Review Instructions**
A "More" button should be visible in the sidebar accordion for private workflows, tools, and notebooks, which should reveal the appropriate "Discover Existing [Workflows/Tools/Notebooks]" button on click.

In terms of the syncing functionality, clicking "Discover Existing" for each of `my-workflows`, `my-tools`, and `my-notebooks` should display synced entries of the correct type.

**Issue**
https://github.com/dockstore/dockstore/issues/5798

**Security**
If there are any concerns that require extra attention from the security team, highlight them here.

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that your code compiles by running `npm run build`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] If this is the first time you're submitting a PR or even if you just need a refresher, consider reviewing our [style guide](https://github.com/dockstore/dockstore/wiki/Dockstore-Frontend-Opinionated-Style-Guide#pr-checklist)
- [x] Do not bypass Angular sanitization (bypassSecurityTrustHtml, etc.), or justify why you need to do so
- [x] If displaying markdown, use the `markdown-wrapper` component, which does extra sanitization
- [x] Do not use cookies, although this may change in the future
- [x] Run `npm audit` and ensure you are not introducing new vulnerabilities
- [x] Do due diligence on new 3rd party libraries, checking for CVEs
- [x] Don't allow user-uploaded images to be served from the Dockstore domain
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead.
- [x] Check whether this PR disables tests. If it legitimately needs to disable a test, create a new ticket to re-enable it in a specific milestone.

### CODE REVIEW

File: cypress/e2e/group2/myworkflows.ts [hunk 1]
```diff
@@ -52,8 +52,15 @@ describe('Dockstore my workflows', () => {
 
     cy.visit('/my-workflows');
     cy.get('[data-cy=myWorkflowsMoreActionButtons]').should('be.visible').click();
+    cy.fixture('myWorkflows.json').then((json) => {
+      cy.intercept('GET', '/api/users/1/workflows', {
+        body: json,
+        statusCode: 200,
+      }).as('getWorkflows');
+    });
```
Comment (ll5zh) [inline cypress/e2e/group2/myworkflows.ts:60]: Since the UI now issues a new GET request for the user's entries after clicking "Discover existing", these stubbed requests were added for the integration tests to work. I suppose the tests' main purpose is to check that the correct endpoint (workflows vs notebooks vs containers/appTools) is being called.

**Review by coverbeck — COMMENTED** [review]
The other option is to change the web service endpoint. I know I was the one to propose making a second request from the browser, but it was intended as more of  a stopgap to try to squeeze in a fix to 1.15, which we ended up not deciding to do.

I left a lot of comments, which are probably confusing. As a minimum, I'd be happy with the code comment suggested in user.service.ts.

File: src/app/shared/user/user.service.ts [hunk 2]
```diff
@@ -50,12 +56,16 @@ export class UserService {
...
+  addUserToWorkflows(userId: number, entryType: EntryType): void {
     this.alertService.start('Adding user to existing workflows and tools on Dockstore');
     this.usersService.addUserToDockstoreWorkflows(userId).subscribe(
       (workflows: Array<Workflow>) => {
         this.alertService.detailedSuccess();
-        this.workflowService.setWorkflows(workflows);
+        if (entryType === EntryType.Tool) {
+          this.mytoolsService.getMyEntries(userId, entryType);
+        } else {
+          this.myWorkflowsService.getMyEntries(userId, entryType);
```
[the remark this answers is not in the document]
Comment (svonworl) [reply src/app/shared/user/user.service.ts:67]: > For the `else` condition, you already have the `workflows` from line 56/62, and you could just do a filter in Typescript instead of making another (relatively) expensive network call.

Not currently possible, because the `addUserToDockstoreWorkflows` endpoint only returns the user's `BioWorkflows`, so there's no `Notebooks` or other stuff to filter for.

I like the two-step "'add' and then 'get'" mechanism.  Sure, there's a little extra latency, but it seems to separate the concerns better and only retrieves entries of the necessary type.  If we stuck with it, I'd propose that we [eventually?] change the `addUserToDockstoreWorkflows` endpoint to not return anything.

**Review by svonworl — APPROVED** [review]

File: src/app/shared/user/user.service.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (coverbeck) [reply src/app/shared/user/user.service.ts:67]: Ah, you're right, I saw the method returns `Workflow` and assumed it would return all subclasses of `Workflow`, but it does only return `BioWorkflow`.

[the remark this answers is not in the document]
Comment (coverbeck) [reply src/app/shared/user/user.service.ts:67]: Followup to my followup, you could just use `workflows` if the entry type is BioWorkflow and not make the extra call. But for symmetry, maybe it's best to keep it as is.

> I'd propose that we [eventually?] change the addUserToDockstoreWorkflows endpoint to not return anything.

Not sure I agree with that one, but we can cross that bridge if we get there.

[the remark this answers is not in the document]
Comment (ll5zh) [reply src/app/shared/user/user.service.ts:67]: > Followup to my followup, you could just use `workflows` if the entry type is BioWorkflow and not make the extra call.

I don't know if this happens for others, but on my end (locally), when the UI uses the `workflows` response instead of `getMyEntries`, clicking "Discover Existing Dockstore Workflows" doesn't display the newly synced workflows right away:

Screenshot from 2024-04-03 16-05-37

If I refresh the page though, all the expected workflows repopulate the page/sidebar. Could be a personal issue... does anyone else encounter something like this? (noting that I had zero private workflows prior to calling the endpoint)

[the remark this answers is not in the document]
Comment (coverbeck) [reply src/app/shared/user/user.service.ts:67]: Do you mean this happens with your fix or without your fix? Or both?

Also, you should be able to see the network response in the browser dev tools; does that look correct?

[the remark this answers is not in the document]
Comment (ll5zh) [reply src/app/shared/user/user.service.ts:67]: For me it happens without the fix, with the original `this.workflowService.setWorkflows(workflows);` line. Network looks okay too (request to /api/users/{id}/workflows, 200 status code, response is the array of `BioWorkflow` entries).

[the remark this answers is not in the document]
Comment (coverbeck) [reply src/app/shared/user/user.service.ts:67]: I'll give it a shot locally when I get a chance, but I may be delayed.

[the remark this answers is not in the document]
Comment (ll5zh) [reply src/app/shared/user/user.service.ts:69]: > on my end (locally), when the UI uses the `workflows` response instead of `getMyEntries`, clicking "Discover Existing Dockstore Workflows" doesn't display the newly synced workflows right away:

(referring to the original code without the fix) Hmm, I think this might be an edge case for users that don't have any private workflows. I tried registering a test workflow before clicking "Discover Existing Dockstore Workflows", and everything appeared to display correctly.I guess it's because, when a user goes from zero to 1+ workflows via syncing, the UI needs to reprocess the workflow to be selected/displayed?)

If this is the case, I think it might be ideal to just always make the second request. With the `getMyEntries` fix, I didn't notice these issues in the UI, even though I had zero private workflows prior to syncing.

**Comment by ll5zh** [comment]
Looks like this line https://github.com/dockstore/dockstore/blob/develop/dockstore-webservice/src/main/java/io/dockstore/webservice/resources/UserResource.java#L1112 is why the endpoint only returns `BioWorkflow` entries:

`return convertWorkflowSummariesToWorkflows(this.bioWorkflowDAO.findUserBioWorkflows(user.getId()), BioWorkflow::new);`

If we were to solve the bug through the web service instead of the UI, the endpoint would need to know the entry type/page that the user is currently on (e.g. return `findUserBioWorkflows` for my-workflows, return `findUserNotebooks` for my-notebooks). I assume my-tools would involve returning an array of workflows combining both `AppTool` and `Container` entries. (Unless the endpoint were to return all workflows, for the UI to filter by entry type? Would that be even more load/network calling?)

> I'd propose that we [eventually?] change the addUserToDockstoreWorkflows endpoint to not return anything.

Personally I think I might agree, since the endpoint's return value seems like a separate call, which doesn't really contribute to the "adding users to Dockstore workflows". Also, when the GET request is issued through the UI (let's say it calls `getMyBioWorkflows` from `getMyEntries`), I believe it [reaches the same return value/functionality](https://github.com/dockstore/dockstore/blob/7347da6a1ecb1c7f8d1385c4d17b273bf1244b21/dockstore-webservice/src/main/java/io/dockstore/webservice/resources/UserResource.java#L628) anyways. So it seems okay to remove the return statement, and let the UI handle the GET separately from the PATCH endpoint. (but agreed that this is probably a future concern)

File: src/app/shared/user/user.service.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (coverbeck) [reply src/app/shared/user/user.service.ts:69]: > I guess it's because, when a user goes from zero to 1+ workflows via syncing, the UI needs to reprocess the workflow to be selected/displayed?

Hmm, not sure, but that's a good theory. But yeah, it's probably best to keep the the code the same for all entry types.

**Review by coverbeck — APPROVED** [review]

**Review by denis-yuen — APPROVED** [review]
